src/componentWithName.js

```diff
@@ -16,7 +16,7 @@ function hasName(name, prop, propName, componentName) {
   }
 
   const { type } = prop;
-  if (getComponentName(type) !== name) {
+  if (!new RegExp(name).test(getComponentName(type))) {
```

Comment:
This should definitely not be overloaded - I'd prefer an entirely separate propType, or, a typeof check, so that you never need `new RegExp` - in other words, the *user* should create the regex, not us.